app/common/renderer/lib/vendor/base.js

```diff
@@ -11,59 +11,122 @@ export class BaseVendor {
-      this._sessionCaps[name] = value;
+  checkInputPropertyPresence(vendorName, propertyList) {
+    const missingProps = [];
+    for (const prop of propertyList) {
+      if (!prop.val) {
+        missingProps.push(i18n.t(prop.name));
+      }
+    }
+    if (missingProps.length > 0) {
+      throw new Error(
+        i18n.t('missingVendorProperties', {
+          vendorName,
+          vendorProps: missingProps.join(', '),
+        }),
+      );
     }
   }
 
   /**
+   * Check validity of the WebDriver URL, which may be required by the vendor
    *
    * @param {string} url
    * @returns {URL}
    */
-  _validateUrl(url) {
+  validateUrl(url) {
     let webdriverUrl;
     try {
       webdriverUrl = new URL(url);
     } catch {
-      throw new Error(`${this._translate('Invalid URL:')} ${webdriverUrl}`);
+      throw new Error(`${i18n.t('Invalid URL:')} ${webdriverUrl}`);
     }
     return webdriverUrl;
   }
+
+  /**
+   * Locally save all class properties required to access the vendor server
+   *
+   * @param {Object} vendor
+   * @param {VendorProperties} vendorProperties
+   */
+  saveProperties(vendor, {host, path, port, https, username, accessKey, headers}) {
+    // It is fine to assign all parameters to 'vendor' values -
+    // they are only saved in Redux and not sent to the actual server
+    this.host = vendor.host = host;
+    this.path = vendor.path = path;
+    this.port = vendor.port = port;
+    this.https = vendor.ssl = https;
+    this.username = username;
+    this.accessKey = accessKey;
+    this.headers = headers;
+  }
+
+  /**
+   *
+   * @param {string} name
+   * @param {any} value
+   * @param {boolean} [merge=true]
+   */
+  updateSessionCap(name, value, merge = true) {
+    const previousValue = this._sessionCaps[name];
+    if (merge && _.isPlainObject(previousValue) && _.isPlainObject(value)) {
+      this._sessionCaps[name] = {
+        ...previousValue,
+        ...value,
+      };
+    } else {
+      this._sessionCaps[name] = value;
+    }
+  }
 }
 
 /**
  * @typedef {Object} VendorProperties
  * @property {string} [host='127.0.0.1'] Server host name
  * @property {number|string} [port=4723] Server port
- * @property {string} [username] Optional auth username for HTTP basic auth.
- * @property {string} [accessKey] Optional auth password for HTTP basic auth.
- * @property {boolean} [https=false] Whether to use https protocol while conecting to the server
+ * @property {string} [username] Optional username for authenticating to vendor service
+ * @property {string} [accessKey] Optional password/access key for authenticating to vendor service
+ * @property {boolean} [https=false] Whether to use https protocol while connecting to the server
  * @property {string} [path='/'] Server pathname
  * @property {Record<string, string>} [headers] Optional server headers
  */
+
+/**
+ * @typedef {Object} InputProperty
+ * @property {string} [name] Property name (used in error messages if no value is provided)
```

Comment:
I assume name is required and cannot be undefined (e.g. square brackets are not needed)